pkg/logger/json.go

```diff
@@ -0,0 +1,249 @@
+	minLevel := l.level
+	if minLevel > level {
+		return
+	}
+
+	t := l.timeFnc()
+
+	l.mutex.Lock()
+	defer l.mutex.Unlock()
+
+	l.logJSON(t, name, level, msg, args...)
+
+	l.writer.Flush(level)
+}
+
+func (l *jsonLogger) logJSON(t time.Time, name string, level LogLevel, msg string, args ...interface{}) {
+	vals := l.getVals(t, name, level, msg)
+
+	if args != nil && len(args) > 0 {
+		for i := 0; i < len(args); i = i + 2 {
+			val := args[i+1]
+			switch sv := val.(type) {
+			case error:
+				switch sv.(type) {
+				case json.Marshaler, encoding.TextMarshaler:
+				default:
+					val = sv.Error()
+				}
+			}
+
+			var key string
+
+			switch st := args[i].(type) {
+			case string:
+				key = st
+			default:
+				key = fmt.Sprintf("%s", st)
+			}
+			vals[key] = val
+		}
+	}
+
+	err := json.NewEncoder(l.writer).Encode(vals)
+	if err != nil {
+		if _, ok := err.(*json.UnsupportedTypeError); ok {
+			vals := l.getVals(t, name, level, msg)
+			vals["warn"] = errInvalidTypeMsg
+
+			json.NewEncoder(l.writer).Encode(vals)
+		}
+	}
+}
+
+func (l jsonLogger) getVals(t time.Time, name string, level LogLevel, msg string) map[string]interface{} {
+	vals := map[string]interface{}{
+		"message":   msg,
+		"timestamp": t.Format(l.timeFormat),
+	}
+
+	if name != "" {
+		vals["module"] = name
+	}
+
+	var levelStr string
+	switch level {
+	case LogError:
+		levelStr = "error"
+	case LogWarn:
+		levelStr = "warn"
+	case LogInfo:
+		levelStr = "info"
+	case LogDebug:
+		levelStr = "debug"
+	default:
+		levelStr = "all"
+	}
+
+	vals["level"] = levelStr
+
+	if pc, file, line, ok := runtime.Caller(defaultCallerOffset + 1); ok {
+		vals["caller"] = fmt.Sprintf("%s:%d", file, line)
+		if ok {
+			f := runtime.FuncForPC(pc)
+			vals["component"] = f.Name()
+		}
+	}
+
+	return vals
+}
+
+// Debugf prints the message and args at DEBUG level
+func (l *jsonLogger) Debugf(msg string, args ...interface{}) {
+	l.log(l.Name(), LogDebug, msg, args...)
+}
+
+// Infof prints the message and args at INFO level
+func (l *jsonLogger) Infof(msg string, args ...interface{}) {
+	l.log(l.Name(), LogInfo, msg, args...)
+}
+
+// Warningf prints the message and args at WARN level
+func (l *jsonLogger) Warningf(msg string, args ...interface{}) {
+	l.log(l.Name(), LogWarn, msg, args...)
+}
+
+// Errorf prints the message and args at ERROR level
+func (l *jsonLogger) Errorf(msg string, args ...interface{}) {
+	l.log(l.Name(), LogError, msg, args...)
+}
+
+// Update the logging level
+func (l *jsonLogger) SetLogLevel(level LogLevel) {
+	l.level = level
+}
+
+// Name returns the loggers name
+func (i *jsonLogger) Name() string {
+	return i.name
+}
+
+// CloneWithLevel returns a shallow copy of the jsonLogger
+func (l *jsonLogger) CloneWithLevel(level LogLevel) Logger {
+	sl := *l
+	sl.level = l.level
+	return &sl
+}
+
+type writer struct {
+	b bytes.Buffer
+	w io.Writer
+}
+
+func newWriter(w io.Writer) *writer {
```

Comment:
You are actually right, it would be much better. Since we have a file writer and I need the io writer specifically to close the file, I am keeping it different (the buffer isn't being used in case of json since the output is written to the io.Writer directly). But I will use the bufio.NewWriter in the later PR once I turn this to an interface, and unify the logger pkg